### ISSUE TITLE
ChannelArbitrator still awaiting contract resolution (Bitcoin TXN not broadcast)

### ISSUE THREAD

**Issue by cyclopes2011** [lightningnetwork/lnd#1071]
### Background

First, my intent was to increase my channel funding across the board so I was starting to close channels. I was unable to close a channel normally and also with --force. I received the "channel not found" error. I then learned about adding the --output_value which did find/close the channel but I forgot to remove the --force option. I believe the remote node was online when this command executed and did not intend to force close a active node's channel.

The force close was executed around April 6 and today is still in the state below.
{
            "channel": {
                "remote_node_pub": "024a2e265cd66066b78a788ae615acdc84b5b0dec9efac36d7ac87513015eaf6ed",
                "channel_point": "433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1",
                "capacity": "125000",
                "local_balance": "123733",
                "remote_balance": "0"
            },
            "closing_txid": "842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae381",
            "limbo_balance": "123733",
            "maturity_height": 0,
            "blocks_til_maturity": 0,
            "recovered_balance": "0",
            "pending_htlcs": [
            ]

The transaction is not found on the Bitcoin explorer (https://blockchain.info/search?search=842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae38). I thought this was due to the contract timeout/wait period due to the --force (uncooperative close). To test, I force closed another channel (with an offline node) and the transaction was broadcast immediately and shows a positive value for "blocks_til_maturity" (which is now decreasing normally).

From lnd.log:
2018-04-10 09:09:48.450 [ERR] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): unable to progress resolver: quitting
2018-04-10 09:11:12.651 [INF] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): starting state=StateWaitingFullResolu
tion
2018-04-10 09:11:12.655 [INF] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): still awaiting contract resolution
2018-04-10 09:11:12.724 [INF] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): relaunching 1 contract resolvers

Apologies but I have overrun the logs after switching to a "debug" level of logging to find more detail so I am not able to provide any further logs of the actual closing. There is nothing in the bitcoin logs for:

bitcoin-cli gettxout "842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae381" 0

To check, the following transaction does appear for the test "force" channel close mentioned earlier:

bitcoin-cli gettxout "1d0d51f27d7f7d0670db8547a76fa39abea55d51d5009ae405db032da0ee781d" 0
{
  "bestblock": "00000000000000000006a1f76c18d2d0445e98991e29e26832955b6b217092fd",
  "confirmations": 122,
  "value": 0.00121923,
  "scriptPubKey": {
    "asm": "0 32cb2dd2957bda0be6b3ba17c93365654840d616e6f9451671e1c69de594b79f",
    "hex": "002032cb2dd2957bda0be6b3ba17c93365654840d616e6f9451671e1c69de594b79f",
    "reqSigs": 1,
    "type": "witness_v0_scripthash",
    "addresses": [
      "bc1qxt9jm55400dqhe4nhgtujvm9v4yyp4skumu529n3u8rfmev5k70sywcyzx"
    ]
  },
  "coinbase": false
}

Is there a mechanism for the LND to rebroadcast a transaction in this state? Is there a manual process to use in order to recover the funds?

### Your environment

* (Installed April 10 prior to submitting this issue)
  - lnd version 0.4.1-beta commit=12cb35...
* Linux steropes 4.13.0-38-generic #43~16.04.1-Ubuntu SMP Wed Mar 14 17:48:43 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
* Bitcoin Core RPC client version v0.16.0.0-g4b4d7eb255
* steropes (0332052748755d74b3395b91eaea78b53bd64b55eb031ccb6a7ae23ac445bb6af7) (Bitcoin mainnet) Functioning normally and have been on mainnet since early March.
* Lowered minrelay fee in bitcoind to 0.00000001 and restarted both bitcoind and LND based on other discussions and in the Slack channels but this appears to only affect future transactions.

### Steps to reproduce

Unable to reproduce--tested with another channel close with --force but with an offline node. This test is proceeding normally as shown by "blocks_til_maturity" decreasing.

### Expected behaviour

(I'm not a developer/coder so this is just my logic as a suggestion.) If the contract resolution or other timeouts expire without resolution maybe the transaction could be rebroadcast. Or, an error condition should be pushed to the user for further action (even if manual during beta phase). Otherwise, you are the experts. :)

### Actual behaviour

Force channel close has resulted in a hung state at "blocks_til_maturity" at 0 value. Transaction was never sent by bitcoind or possibly LND. No re-transmit mechanism or error to user exists for this case. See information above.

### Next steps
Please let me know if additional detail is required. The LND logs are probably a no-go for any historical information but the bitcoind logs should be fine.

Thank you for all your efforts and expertise.

**Comment by Roasbeef** [comment]
So I see the closing transaction here: https://www.smartbit.com.au/tx/306842fa2172d2142909d2f483390cb5289708433b1d6b52809be1e1320e8a60

Note that if you force closed a channel, but didn't have any active funds, then we won't need to track it any longer.

**Comment by Roasbeef** [comment]
We'll auto broadcast transactions at this point. Perhaps the remote party also force closed the transaction, and it was the alternative transaction that was chosen for block inclusion. If this was the case, a pending PR will resolve this issue.

**Comment by Roasbeef** [comment]
Looking at the `txid`'s you gave above, the dates don't match up. The channel link was fully resolved on chain at April 4th, yet your claim is that you force closed on April 6th?

**Comment by cyclopes2011** [comment]
Isn't 306842fa2172d2142909d2f483390cb5289708433b1d6b52809be1e1320e8a60 the opening transaction to create the channel? The funding amount was 0.00125 BTC. The closing transaction would be less than the 0.00125 BTC minus on-chain fees--shown in the output in the Background section above.

I believe I did the force close "around April 6" so, yes, I could be off a day or two. :)

I pulled that transaction from my bitcoind debug.log:

 bitcoin-cli gettxout "306842fa2172d2142909d2f483390cb5289708433b1d6b52809be1e1320e8a60" 0
{
  "bestblock": "00000000000000000002600e6fcc926667227fe92f3ad593d8e46644176883ae",
  "confirmations": 920,
  "value": 0.00124276,
  "scriptPubKey": {
    "asm": "0 412dc99eeab3e5b134ed34b97a4be5190b0d1697",
    "hex": "0014412dc99eeab3e5b134ed34b97a4be5190b0d1697",
    "reqSigs": 1,
    "type": "witness_v0_keyhash",
    "addresses": [
      "bc1qgykun8h2k0jmzd8dxjuh5jl9ry9s695h9ujzph"
    ]
  },
  "coinbase": false
}

Thank you for looking at this. Please let me know if there is anything further. I will wait on the PR and maintain this node in its current state until resolved. Thank you again. :)

**Comment by Roasbeef** [comment]
You sent a chan point of: 
```
{
"channel": {
"remote_node_pub": "024a2e265cd66066b78a788ae615acdc84b5b0dec9efac36d7ac87513015eaf6ed",
"channel_point": "433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1",
```

That means the opening transaction txid is `433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289`.

**Comment by Roasbeef** [comment]
From examining the chain: all funds were fully swept.

**Comment by cyclopes2011** [comment]
Output as of now:
lncli pendingchannels | more
{
    "total_limbo_balance": "123733",
    "pending_open_channels": [
    ],
    "pending_closing_channels": [
    ],
    "pending_force_closing_channels": [
        {
            "channel": {
                "remote_node_pub": "024a2e265cd66066b78a788ae615acdc84b5b0dec9efac36d7ac87513015eaf6ed",
                "channel_point": "433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1",
                "capacity": "125000",
                "local_balance": "123733",
                "remote_balance": "0"
            },
            "closing_txid": "842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae381",
            "limbo_balance": "123733",
            "maturity_height": 0,
            "blocks_til_maturity": 0,
            "recovered_balance": "0",
            "pending_htlcs": [
            ]
        }

**Comment by Roasbeef** [comment]
What I'm saying is the _other_ commitment transaction was confirmed, and the funds were swept (as both outputs are spent on-chain), but it's a cosmetic bug that the RPC itself wasn't properly updated. 

One restarts, if you set `CNCT=trace` as a debug level, what logs appear (if you grep for `CNCT`)?

**Comment by Roasbeef** [comment]
This is the commitment transaction: https://www.smartbit.com.au/tx/433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289

Observe that both outputs are spent.

**Comment by cyclopes2011** [comment]
If this is cosmetic, I don't want to bother everyone with a low priority item. I will continue to provide detail if needed. :)

With the CNCT=trace option and grep for the commitment transaction:

cat lnd.log | grep 433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289

2018-04-10 18:52:43.282 [ERR] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): unable to progress resolver: quitting
2018-04-10 18:52:43.285 [TRC] CNCT: Attempting to stop ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1)
2018-04-10 18:52:43.285 [DBG] CNCT: Stopping ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1)
2018-04-10 18:53:10.383 [INF] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): starting state=StateWaitingFullResolution
2018-04-10 18:53:10.386 [INF] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): still awaiting contract resolution
2018-04-10 18:53:10.480 [INF] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): relaunching 1 contract resolvers

**Comment by cyclopes2011** [comment]
Full output...

[cnct_issue101.log](https://github.com/lightningnetwork/lnd/files/1896719/cnct_issue101.log)

**Comment by halseth** [comment]
Should be fixed by #1088?

**Comment by cyclopes2011** [comment]
After updating from master, I believe I successfully installed the PR 1088 since it is not merged into the master as of now. If I recall correctly, I did a 'git checkout 1088' and a 'make && make install.' (This is my first experience with git.)

```
lncli --version
lnd version 0.4.1-beta commit=aaa8fa33b1ec43f5b993c986826d5eb3e179066
```
(which matches the commit for 1088)

Currently:
```
lncli pendingchannels
...
"pending_force_closing_channels": [
        {
            "channel": {
                "remote_node_pub": "024a2e265cd66066b78a788ae615acdc84b5b0dec9efac36d7ac87513015eaf6ed",
                "channel_point": "433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1",
                "capacity": "125000",
                "local_balance": "123733",
                "remote_balance": "0"
            },
            "closing_txid": "842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae381",
            "limbo_balance": "123733",
            "maturity_height": 0,
            "blocks_til_maturity": 0,
            "recovered_balance": "0",
            "pending_htlcs": [
            ]
```

And logs (with CNCT trace on):

```
2018-04-13 10:29:13.275 [ERR] UTXN: Notification chan closed, can't advance output 842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae381:0
2018-04-13 10:29:13.275 [ERR] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): unable to progress resolver: quitting
2018-04-13 10:29:35.420 [INF] UTXN: Commitment outpoint 842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae381:0 registered for confirmation notification.
2018-04-13 10:29:35.420 [INF] NTFN: New confirmations subscription: txid=842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae381, numconfs=1
2018-04-13 10:29:38.392 [INF] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): starting state=StateWaitingFullResolution
2018-04-13 10:29:38.395 [INF] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): still awaiting contract resolution
2018-04-13 10:29:38.461 [INF] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): relaunching 1 contract resolvers
2018-04-13 10:29:38.463 [INF] NTFN: New confirmations subscription: txid=842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae381, numconfs=1
2018-04-13 10:34:30.403 [ERR] UTXN: Notification chan closed, can't advance output 842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae381:0
2018-04-13 10:34:30.403 [ERR] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): unable to progress resolver: quitting
2018-04-13 10:35:58.566 [INF] UTXN: Commitment outpoint 842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae381:0 registered for confirmation notification.
2018-04-13 10:35:58.567 [INF] NTFN: New confirmations subscription: txid=842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae381, numconfs=1
2018-04-13 10:36:01.404 [DBG] CNCT: Starting ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1), htlc_set=(contractcourt.htlcSet) {
2018-04-13 10:36:01.404 [INF] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): starting state=StateWaitingFullResolution
2018-04-13 10:36:01.407 [TRC] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): attempting state step with trigger=chainTrigger
2018-04-13 10:36:01.408 [INF] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): still awaiting contract resolution
2018-04-13 10:36:01.474 [TRC] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): next_state=StateWaitingFullResolution
2018-04-13 10:36:01.474 [TRC] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): terminating at state=StateWaitingFullResolution
2018-04-13 10:36:01.475 [INF] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): relaunching 1 contract resolvers
2018-04-13 10:36:01.476 [DBG] CNCT: ChannelArbitrator(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): attempting to resolve *contractcourt.commitSweepResolver
2018-04-13 10:36:01.477 [DBG] CNCT: *contractcourt.commitSweepResolver(433cf78c90da64870ffa918e19b52761eae463855f9f56bff1374613aa46f289:1): waiting for commit tx to confirm
2018-04-13 10:36:01.477 [INF] NTFN: New confirmations subscription: txid=842ca6090873a685ca47dd736b65445bb8a894edceeb69a738d7514605aae381, numconfs=1
```

I also noticed this but not sure if it is relevant:
`2018-04-13 10:36:22.779 [ERR] SRVR: unable to get channel links: unable to locate channel link by destination hop id 024a2e265cd66066b78a788ae615acdc84b5b0dec9efac36d7ac87513015eaf6ed`

I will continue to monitor and update. Please let me know what additional information you need.

**Comment by Roasbeef** [comment]
Worth noting that _all funds_ have been swept from that channel, so your wallet has the proper balance. The issue is that we attempted to play one commit, while the other confirmed. This will be fixed by #1017.

**Comment by cyclopes2011** [comment]
Thank you. I will continue to monitor and return to the master version--likely after #1017 is merged.

**Comment by Roasbeef** [comment]
Both outputs on that commitment have been swept. This means that all the coins in the channel has been fully recovered by `lnd`. The channel still showing up in the `pendingchannels` call is more or less, a cosmetic blemish. The root cause has been fixed in master, but if your internal state acted before this fix, then it will remain in that state. It's important to note it's a slight DB inconsitency (that's been fixed now), and the coins have been recovered by the wallet. 

Basically, the resolver is waiting for a transaction that will never confirm, since it's the case that _another_ commitment transaction confirmed instead.

**Comment by Roasbeef** [comment]
We'll be adding a "forget channels" debug RPC soon-ish. This will let users clear up state like this due to fixed bugs.

**Comment by cyclopes2011** [comment]
Understood...just providing follow-up. Thank you again!
